### ISSUE TITLE
omode mapping incorrectly used when playing a macro

### ISSUE THREAD

**Issue by nall** [vim/vim#6350]
**Describe the bug**
An omode mapping is errantly used when playing a macro

**To Reproduce**
1. Create a vimrc:
```
function! MyFunc()
    echom "hi from func"
endfunction

onoremap , :<c-u>call MyFunc()<CR>
```

2. Run vim with `vim -u <rcfile from above>` and in the new buffer and type: `foo([blah])`
3. In normal mode, starting on `f`, record a macro: `ct[,bar<Esc>`
4. Undo that change and then play the macro on the text in step 2 above. When **playing** the macro I instead see a result of `:<c-u>call MyFunc()<CR>bar[blah])`.

I expect vim to be in insert mode when it processes the `,` in the macro, but it seems that character is being interpreted in the context of operator pending mode.

**Expected behavior**
I expect the playing the macro to result in `,bar[blah])`. In fact, this is what happens when **recording** the macro.

**Environment (please complete the following information):**
  - Vim version: tried on 8.2.1050 (a reddit user reproduced on 8.2.539 and 8.0.1365)
  - OS: macOS 10.15
  - Terminal: iTerm2

**Additional context**
If I `:ounmap ,`, the macro plays as expected.

**Comment by lacygoill** [comment]
Possible duplicate of https://github.com/vim/vim/issues/3021 and https://github.com/vim/vim/issues/3678.

**Comment by nall** [comment]
FYI per neovim suggestion, I created [an issue](https://github.com/neovim/neovim/issues/12551) for the bug there as well and supplied a link to this issue.

**Comment by lacygoill** [comment]
And there is [a todo item](https://github.com/vim/vim/blob/ab19d495fd880b25a38d58cbeb5b21e4d0ee5835/runtime/doc/todo.txt#L635) for this issue:

> nvo-mode mapping works on Windows, not on Linux. (#3678)

**Comment by zeertzjq** [comment]
There is a related note in docs:
```
Keep in mind that the expression may be evaluated when looking for
typeahead, before the previous command has been executed.  For example: >
	func StoreColumn()
	  let g:column = col('.')
	  return 'x'
	endfunc
	nnoremap <expr> x StoreColumn()
	nmap ! f!x
You will notice that g:column has the value from before executing "f!",
because "x" is evaluated before "f!" is executed.
This can be solved by inserting <Ignore> before the character that is
expression-mapped: >
	nmap ! f!<Ignore>x
```
But that workaround doesn't work in this case.

**Comment by chrisbra** [comment]
yeah, but it's not an expression mapping. Actually wondering why the `operator` mapping is triggered at all 🤔

**Comment by chrisbra** [comment]
Ah thanks, Looks like we are running into this situation, but now with the opposite problem as described here:

https://github.com/vim/vim/blob/4c0089d696b8d1d5dc40568f25ea5738fa5bbffb/src/normal.c#L548-L550

**Comment by zeertzjq** [comment]
I think this will fix the problem:
```diff
diff --git a/src/getchar.c b/src/getchar.c
index c5ccdf2e0..a807b4551 100644
--- a/src/getchar.c
+++ b/src/getchar.c
@@ -1338,6 +1338,17 @@ gotchars(char_u *chars, int len)
     ++maptick;
 }
 
+    void
+gotchars_nop(void)
+{
+    char_u nop_buf[3];
+
+    nop_buf[0] = K_SPECIAL;
+    nop_buf[1] = KS_EXTRA;
+    nop_buf[2] = KE_NOP;
+    gotchars(nop_buf, 3);
+}
+
 /*
  * Undo the last gotchars() for "len" bytes.  To be used when putting a typed
  * character back into the typeahead buffer, thus gotchars() will be called
@@ -3656,14 +3667,9 @@ vgetorpeek(int advance)
 #endif
     if (timedout && c == ESC)
     {
-	char_u nop_buf[3];
-
 	// When recording there will be no timeout.  Add a <Nop> after the ESC
 	// to avoid that it forms a key code with following characters.
-	nop_buf[0] = K_SPECIAL;
-	nop_buf[1] = KS_EXTRA;
-	nop_buf[2] = KE_NOP;
-	gotchars(nop_buf, 3);
+	gotchars_nop();
     }
 
     --vgetc_busy;
diff --git a/src/normal.c b/src/normal.c
index 4004d4204..2fe64fdea 100644
--- a/src/normal.c
+++ b/src/normal.c
@@ -568,6 +568,8 @@ normal_cmd_get_more_chars(
     --no_mapping;
     --allow_keys;
 
+    gotchars_nop();
+
     return idx;
 }
 
diff --git a/src/proto/getchar.pro b/src/proto/getchar.pro
index 01746cf70..7e05870a9 100644
--- a/src/proto/getchar.pro
+++ b/src/proto/getchar.pro
@@ -30,6 +30,7 @@ int typebuf_changed(int tb_change_cnt);
 int typebuf_typed(void);
 int typebuf_maplen(void);
 void del_typebuf(int len, int offset);
+void gotchars_nop(void);
 void ungetchars(int len);
 int save_typebuf(void);
 void save_typeahead(tasave_T *tp);
```

I'll create a PR soon.

**Comment by AndrewRadev** [comment]
We got an issue about this in vim-ruby as well: https://github.com/vim-ruby/vim-ruby/issues/455. I can confirm that the PR from @zeertzjq seems to fix the macro problem.

That said, the problem also happens if you just use `normal` with a `cf` mapping, and it still happens on his branch as well. As I describe in that issue, I can put the following in a file called test.vim:

``` vim
" One two targeT

onoremap <buffer> fo BUG
```

Putting the cursor in the beginning of `two` and executing `normal cfTfour` results, as expected, in `" One four`. Sourcing the file with `:source %` and then executing `normal cfTfour` again results in `" One BUGur`.

**Comment by zeertzjq** [comment]
This behavior was introduced in [patch 7.3.374](https://github.com/vim/vim/commit/4f88062c556e52368d65991f5ae74cf0f8b17f2d), and it's hard to infer when the problem described in that patch actually can happen. (Maybe inside screen/tmux?)

**Comment by chrisbra** [comment]
Seems, the issue is mentioned here: https://groups.google.com/g/vim_dev/c/cVGmv0RVOP0/m/AU8EInWAk8YJ 

Can you please check, if this problem occurs after #12708 would be merged?

**Comment by zeertzjq** [comment]
Ah, it's specifically referring to `:normal`. Then I think that patch cannot be reverted, and `<Ignore>` can be used to achieve the old behavior in `:normal`.

**Comment by zeertzjq** [comment]
PR #12708 only changes macro recording behavior, it doesn't affect `:normal`.

For `:normal`, before patch 7.4.374 mappings never work immediately after character find, while after that patch mappings may be applied in the wrong mode. Neither is completely "correct", and `<Ignore>` may be used to achieve the "correct" behavior in either case.

**Comment by AndrewRadev** [comment]
@zeertzjq @chrisbra Yes, I see that it's a different issue, and I can confirm that adding `<Ignore>` in the `normal` invocation avoids the problem.

**Comment by chrisbra** [comment]
so are we good to have #12708 merged?
